File: src/main/java/ru/sfedu/teamselection/repository/StudentRepository.java
package ru.sfedu.teamselection.repository;

import java.time.LocalDate;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.JpaSpecificationExecutor;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;
import ru.sfedu.teamselection.domain.Student;


@Repository
public interface StudentRepository extends JpaRepository<Student, Long>, JpaSpecificationExecutor<Student> {

    Student findByUserId(Long userId);

    boolean existsByUserId(Long userId);

    @Modifying(clearAutomatically = true)
    @Query("""
        UPDATE Student s
        SET s.isCaptain = false, s.currentTeam = null, hasTeam = false
        WHERE s.currentTrack IS NOT NULL
        AND s.currentTrack.endDate < :currentDate
    """)
    int deactivateCaptainsWithExpiredTracks(LocalDate currentDate);
}

File: src/main/kotlin/ru/sfedu/teamselection/repository/StudentRepository.kt
package ru.sfedu.teamselection.repository

import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.transaction.annotation.Transactional
import ru.sfedu.teamselection.domain.Student

@Transactional
interface StudentRepository: JpaRepository<Student, Long> {

    fun findByEmail(email: String): Student?
    fun findAllByCaptainAndTrackId(flag: Boolean, id: Long): MutableList<Student>
    fun findStudentByStatusAndTrackId(flag: Boolean, trackId: Long): MutableList<Student>
    fun findStudentByTagsInAndTrackId(tags: List<String>, trackId: Long): MutableList<Student>
    fun findAllByIdIn(ids: List<Long>): MutableList<Student>
    fun findAllByTrackId(id: Long): List<Student>
}